### PR TITLE
Implement multi-value charmaps

### LINKED ISSUE #97: Widechar charmap support
Now RGBDS can only supprot

> charmap "C", $01

Can RGBDS support widechar charmap like

> charmap "C", $01F0

or

> charmap "C", $01F0, 2

?

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #97

This implementation follows @ISSOtm's suggested approach ("we need charmaps to map not to a byte, not to a u32, not to a byte array, but to a u32 array"). That way, you can do `charmap "X", $abcd` if you always plan to use it in a word-size context (i.e. `dw "X"`), or do `charmap "X", $ab, $cd` if you want `db "X"` to work without truncating.

I ended up keeping the ability to treat a multi-byte character as a number, because it's equivalent to treating a multi-character string as a number, which we already allowed (with warnings). This is confusing in some situations, but it already was:

```asm
charmap "A", 1
charmap "B", 2
charmap "C", 3
charmap "D", 4

assert ("ABCD") == $01020304 ; big-endian

; these don't match!?
db "ABCD"   ; 01 02 03 04, character order
dl ("ABCD") ; 04 03 02 01, little-endian
```

### CODE REVIEW

File: src/asm/charmap.cpp [hunk 1]
```diff
@@ -16,10 +17,11 @@
 // Essentially a tree, where each nodes stores a single character's worth of info:
 // whether there exists a mapping that ends at the current character,
 struct CharmapNode {
-	bool isTerminal; // Whether there exists a mapping that ends here
-	uint8_t value;   // If the above is true, its corresponding value
+	std::vector<int32_t> value;   // The mapped value, if there exists a mapping that ends here
```
Comment (ISSOtm) [inline src/asm/charmap.cpp:20]: I miss C's flexible array member... u_u

Comment (Rangi42) [reply src/asm/charmap.cpp:20]: Yeah, I wish this had been a two dozenth realloc'ed-array implementation too :P

File: src/asm/parser.y [hunk 2]
```diff
@@ -2374,26 +2378,37 @@ void yy::parser::error(std::string const &str) {
...
 
-static uint32_t str2int2(std::vector<uint8_t> const &s) {
+static uint32_t str2int2(std::vector<int32_t> const &s) {
 	uint32_t length = s.size();
 
+	if (length == 1) {
+		// The string is a single character with a single value,
+		// which can be used directly as a number.
+		return (uint32_t)s[0];
+	}
```
Comment (Rangi42) [inline src/asm/parser.y:2388]: This single-char single-unit behavior works for the usual cases (`ld a, "X"` where `charmap "X", $42`) as well as allowing FourCC-style codes (`ld a, "RIFF"` where `charmap "RIFF", $52_49_46_46`).

I think this means we can deprecate treating longer strings as numbers. So these two would give a warning:

- `ld a, "ABC"` where `"A"`, `"B"`, and `"C"` are each singe-byte charmaps
- `ld a, "S"` where `charmap "S", 1, 2, 3`

@ISSOtm Should we do that?

Comment (Rangi42) [reply src/asm/parser.y:2388]: Note, if we deprecate it I'd like to do that as a separate PR.

Comment (ISSOtm) [reply src/asm/parser.y:2388]: Yes, please do that.

**Review by ISSOtm — APPROVED** [review]
Thank you very much!

**Pull request merged**